src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -40,20 +58,29 @@ typedef struct alloc_elem {
 
 static struct {
     alloc_elem_t *head, *tail;
-} allocq = {
+} allocq_dram = {
+0}, allocq_hbm = {
 0};
 
 static int check_alloc(MPIDU_shm_seg_t * memory, MPIDU_shm_barrier_t * barrier,
                        int num_local, int local_rank);
 
-#define ALLOCQ_HEAD() GENERIC_Q_HEAD(allocq)
-#define ALLOCQ_EMPTY() GENERIC_Q_EMPTY(allocq)
-#define ALLOCQ_ENQUEUE(ep) GENERIC_Q_ENQUEUE(&allocq, ep, next)
-#define ALLOCQ_DEQUEUE(epp) GENERIC_Q_DEQUEUE(&allocq, epp, next)
-
 #define ROUND_UP_8(x) (((x) + (size_t)7) & ~(size_t)7)  /* rounds up to multiple of 8 */
 
-static size_t segment_len = 0;
+#ifdef HAVE_HWLOC
+#define SET_SHM_OBJ_BINDING(base_addr,len) do {                            \
```

Comment:
@pavanbalaji I hope I have addressed your point (I had to do some more invasive changes tough). I haven't tested this on knl yet. Waiting for JLSE to come back online.